pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/IndexCreatorProviders.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.pinot.segment.spi.index.creator.CombinedInvertedIndexCreator;
+import org.apache.pinot.segment.spi.index.creator.DictionaryBasedInvertedIndexCreator;
+import org.apache.pinot.segment.spi.index.creator.ForwardIndexCreator;
+import org.apache.pinot.segment.spi.index.creator.GeoSpatialIndexCreator;
+import org.apache.pinot.segment.spi.index.creator.JsonIndexCreator;
+import org.apache.pinot.segment.spi.index.creator.TextIndexCreator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Plugin registration point to allow index creation logic to be swapped out.
+ * Plugins should not reimplement Pinot's default index creation logic.
+ * Users provide an override to Pinot's index creation logic. This is simplified
+ * by extending {@see IndexCreatorProviders.Default}
+ */
+public final class IndexCreatorProviders {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(IndexCreatorProviders.class);
+
+  static final IndexCreatorProvider DEFAULT = defaultProvider();
+  private static final AtomicReference<IndexCreatorProvider> REGISTRATION = new AtomicReference<>(DEFAULT);
+
+  private IndexCreatorProviders() {
+  }
+
+  /**
+   * The caller provides a decorator to wrap the default provider, which allows plugins to create
+   * a delegation chain.
+   * @param provider index creation override
+   * @return true if this is the first invocation and the provider has not yet been used.
+   */
+  public static boolean registerProvider(IndexCreatorProvider provider) {
+    return REGISTRATION.compareAndSet(DEFAULT, provider);
+  }
+
+  /**
+   * Obtain the registered index creator provider. If the user has provided an override, then it will be used instead.
+   * If the user has not provided an override yet, then this action will prevent them from doing so.
+   * @return the global index provision logic.
+   */
+  public static IndexCreatorProvider getIndexCreatorProvider() {
+    return Holder.PROVIDER;
+  }
+
+  private static final class Holder {
+    public static final IndexCreatorProvider PROVIDER = REGISTRATION.get();
+  }
+
+  private static IndexCreatorProvider defaultProvider() {
+    // use MethodHandle to break circular dependency and keep implementation details encapsulated within
+    // pinot-segment-local
+    String className = "org.apache.pinot.segment.local.segment.creator.impl.DefaultIndexCreatorProvider";
+    try {
+      Class<?> clazz = Class.forName(className, false, IndexCreatorProviders.class.getClassLoader());
+      return (IndexCreatorProvider) MethodHandles.publicLookup()
+          .findConstructor(clazz, MethodType.methodType(void.class)).invoke();
+    } catch (Throwable implausible) {
+      LOGGER.error("could not construct MethodHandle for {}", className,
+          implausible);
+      throw new IllegalStateException(implausible);
+    }
+  }
+
+  /**
+   * Extend this class to override index creation
+   */
+  public static class Default implements IndexCreatorProvider {
```

Comment:
IIUC, this is the one that allows to mix ? So for example, someone might want to use existing forward index, dictionary and everything else but might want to plugin a different inverted index ? So, this can be sub-classed and newInvertedIndexCreator can be overridden to get custom behavior. The sub-class can then be registered ? Is this flow correct ? 

Also, can we try to add one test (or may not be needed if example usage pattern is going to be covered in docs later on) ?

Reply:
Also I wonder why `DefaultIndexCreatorProvider` is final class because in theory user can just sub-class it and register ?

Reply:
`DefaultIndexCreatorProvider` is in a core module which an external plugin should not have a dependency on, it has access to all kinds of internal APIs which should not be exposed via SPI. The user of the SPI doesn't need a dependency on core modules, just the SPI, and they just need to extend `Default` without knowing what the defaults are. 

For example usage, let's say the user implements a new kind of inverted index only for `INT` columns, for the sake of example, but wants to use the defaults otherwise. They would subclass Default:


```java
public class MyIndexCreatorProvider extends IndexCreatorProviders.Default {
  @Override
  public DictionaryBasedInvertedIndexCreator newInvertedIndexProvider(IndexCreationContext context) {
    if (context.getColumnMetadata().getDataType() == INT) {
      return new MyCustomInvertedIndexCreator(context.getColumnMetadata());
    }
    return super.newInvertedIndexProvider(context)
  }
}
...
IndexCreatorProviders.registerProvider(new MyPluginProvider());
```

The user doesn't care about forward indexes so doesn't implement the method. The idea behind an inheritance based API is that new index types can be added and old extensions will be both source- and binary-compatible.

Reply:
Added a test which should make the mechanism (and motivation) fairly clear: https://github.com/apache/pinot/pull/7885/commits/3ad1746f94f2c82a1347dc084beba25d5fb8e95a